### ISSUE TITLE
Redis Sentinel Server with Stack Exchange Redis Client in a Custom Application Timeout

### ISSUE THREAD

**Issue by ccaedeacon** [StackExchange/StackExchange.Redis#964]
Our Developers have built and launched a custom application running in an IIS application pool utilizing REDIS Sentinel for session management among of pool of 19 Web servers behind a load balancer.

At various times, the IIS server(s) will stop servicing requests for all clients (freeze) - trying to access the webserver will spin your browser during this hang.  This usually appears to be at times that the application is ramping up and consuming more connections and once it stablizes, IIS seems to be stable.

This is what we see in the logs:

2018-09-18 07:54:04.499 -04:00 [ERR] Connection id "0HLGQLD0SASUF", Request id "0HLGQLD0SASUF:00000001": An unhandled exception was thrown by the application.
StackExchange.Redis.RedisTimeoutException: Timeout performing EVAL, inst: 2, queue: 47, qu: 0, qs: 47, qc: 0, wr: 0, wq: 0, in: 20275, ar: 0, clientName: PA-LMSWEB25, serverEndpoint: Unspecified/redis-11000.prodredis.myapplication.com:11000, keyHashSlot: 13540 (Please take a look at this article for some common client-side issues that can cause timeouts: http://stackexchange.github.io/StackExchange.Redis/Timeouts)
   at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\code\StackExchange.Redis\StackExchange.Redis\StackExchange\Redis\ConnectionMultiplexer.cs:line 2120
   at StackExchange.Redis.RedisBase.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\code\StackExchange.Redis\StackExchange.Redis\StackExchange\Redis\RedisBase.cs:line 81
   at StackExchange.Redis.RedisDatabase.ScriptEvaluate(String script, RedisKey[] keys, RedisValue[] values, CommandFlags flags) in c:\code\StackExchange.Redis\StackExchange.Redis\StackExchange\Redis\RedisDatabase.cs:line 1052
   at Microsoft.Extensions.Caching.Redis.RedisExtensions.HashMemberGet(IDatabase cache, String key, String[] members)
   at Microsoft.Extensions.Caching.Redis.RedisCache.GetAndRefresh(String key, Boolean getData)
   at LMS.Presentation.Web.Stores.RedisCacheTicketStore.RetrieveAsync(String key) in C:\Program Files (x86)\Jenkins\workspace\LMS-Production\dotnet\LMS\Presentation\Web\Stores\RedisCacheTicketStore.cs:line 91
   at Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler.<ReadCookieTicket>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler.<HandleAuthenticateAsync>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.<AuthenticateAsync>d__47.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Authentication.AuthenticationService.<AuthenticateAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.<Invoke>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LMS.Presentation.Core.Middlewares.SecurityHeaders.SecurityHeadersMiddleware.<Invoke>d__3.MoveNext() in C:\Program Files (x86)\Jenkins\workspace\LMS-Production\dotnet\LMS\Presentation\Web\Middleware\SecurityHeaders\SecurityHeadersMiddleware.cs:line 64
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.IISIntegration.IISMiddleware.<Invoke>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Hosting.Internal.RequestServicesContainerMiddleware.<Invoke>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Frame`1.<ProcessRequestsAsync>d__2.MoveNext()

I come from the operations side, so everything I can see about the server and network environment is indicating no connectivity or latency problems.

We have wiresharked this, we have cross referenced logs from IIS and logs from the Application Pool, system event logs, redis sentinel logs - nothing is leading to any conclusions.

Regards,

**Comment by mgravell** [comment]
The large "in" there tells me this is a problem in the reader. We have
*completely* re-architected the reader (and writer - everything "network",
basically) in 2.0, specifically to target recurring problems like this -
see https://github.com/StackExchange/StackExchange.Redis/issues/871

From the error message, I *think* you are using v1, in which case my advice
is: "try V2".

(Minor caveat: we're getting some reports of binding redirect problems in
the current build which we have a fix for - I just need to get it deployed,
which should be in the next few hours if our CI server starts cooperating)

**Comment by ccaedeacon** [comment]
Marc,

Thank you so much for the reply.  Chatted with a developer on our side, this was his reply:

"We don't explicitly define what version of StackExchange.Redis we are using, it's imported by the necessary packages. That said, I will be sure to check what version they are using and see what options we have. Thanks for looking into this."

In this case, could you please guide us to how to determine what version we may be using?

Kind regards

**Comment by mgravell** [comment]
If you have an intermediate dependency: you shouldn't update to v2 until
the other tool you're using does. Quite possibly this is aspnetcore.all (or
something like that). In which case: definitely wait for them to update.
But: V2 is how we have addressed this type of problem.

**Comment by NickCraver** [comment]
The issue with v1 was general instability in scenarios we didn't see at Stack Overflow or were generally limited from seeing due to the underlying framework bits used. A _lot_ of those bits were overhauled, new primitives exposed, and that's why v2 was mostly a rewrite of the networking layer. It's easier to debug (can cram much more available info in the exception messages, as an example), easier to work around issues, and much more cross-platform friendly.

In short, there's nothing we can do for v1.x users than suggesting an increased timeout when that is a viable workaround. The reasons for a v2 rewrite and months of time in it were because of this...it was _needed_.

I apologize some old threads got dropped here as we prioritized hours were we had them, but the result/recommendation is the same: please upgrade off v1.2.6 for anyone who can. I sincerely hope that's happened by now and that you're in a good spot.
